### ISSUE TITLE
Wasm I/0 Support

### ISSUE THREAD

**Issue by nxtexe** [WebAssembly/design#1323]
Hi all, I've been using emscripten for less than a week now and I've been intrigued at the possibilities of compiled binaries on the web with the web being the most ubiquitous platform to target. On investing deeper into the mystery of web assembly however I realized something that is quite odd. Why doesn't web assembly currently support input and output to the DOM? If this question is a bit stupid please point it out so this issue can be closed immediately, I'm just at a loss trying to figure out if it's impossible or just hasn't been done yet. As you can imagine a standard input and output interface would be amazing since wasm code could truly reach peak performance as it would no longer have to be importing and exporting functionality between two languages. I've also heard that javascript and wasm can access a common block of memory in order to share data more quickly. I was wary of said approach however since I wouldn't want the two languages stepping on each other's toes overwriting data that one still has use for. Let me know guys. Thanks!

**Comment by Pauan** [comment]
It is possible to access the DOM from Wasm (both [emscripten](https://emscripten.org/docs/api_reference/html5.h.html) and [Rust](https://crates.io/crates/web-sys) can do it).

Currently this requires a heap in JS to convert things into pointers, but after [reference types](https://github.com/WebAssembly/reference-types) are standardized and widely implemented we can bind directly to JS objects. This will make things faster.

So, in other words, Wasm can currently do everything that JS can do (by using some auto-generated JS glue code), and in the near future it will become even faster and easier.

Also, for I/O you should check out [WASI](https://wasi.dev/).

**Comment by nxtexe** [comment]
Thanks a lot and I'll definitely check wasi out. You guys are doing
excellent work keep it up.

**Comment by Xyncgas** [comment]
what about electron, can we have native filesystem to play with, like what they say toys that electron can let webassembly have and do stuffs to the user's actual filesystem while electron handles the CSP

**Comment by kmiller68** [comment]
> what about electron, can we have native filesystem to play with, like what they say toys that electron can let webassembly have and do stuffs to the user's actual filesystem while electron handles the CSP

That's not really a WASM feature. That's something to bring up with the Electron folks, it's possible that's an API they can add.

WASM itself, fundamentally, only deals with compute and has no concept of I/O beyond importing functions from its host environment (e.g. HTML). I/O will only ever be possible because the host provides it as a function that can be imported.

I'm gonna close this because it's out of scope of the WASM project as it stands today. Feel free to reopen or continue commenting if there are any other questions/concerns.

**Comment by Xyncgas** [comment]
so web-assembly DOES not imply restrictions on accessing/using the client's native filesystem, but instead leaving the implementation that uses web-assembly to decide right? And so if I use "fstream.open()" in c++, web-assembly actually don't do anything besides sending a signal to the client, and in this case browsers will try to use a virtual system with files preloaded/embedded during wasm compilation while it's possible for electron to handle this signal by actually providing native filesystem support to the fstream call from c++?

**Comment by kmiller68** [comment]
> so web-assembly DOES not imply restrictions on accessing/using the client's native filesystem, but instead leaving the implementation that uses web-assembly to decide right? 

Correct, that is up to the host, in much the same way it is for platforms that embed JS.

> And so if I use "fstream.open()" in c++, web-assembly actually don't do anything besides sending a signal to the client, and in this case browsers will try to use a virtual system with files preloaded/embedded during wasm compilation while it's possible for electron to handle this signal by actually providing native filesystem support to the fstream call from c++?

Browsers don't have any concept of `fstream.open()`. it's up to things like emscripten to re-implement the C++ standard library on top of the existing JS APIs the browser provides today. Then forward those functions to your compiled WASM code.

**Comment by Xyncgas** [comment]
Thanks.
